ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,358 @@
+    :gen_tcp.send(sock, Frames.request(0, next_seq, encoded_payload))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_cast(_, %State{sequence: :go_away}=state),
+    do: {:noreply, state}
+
+  def handle_cast({:push, payload}, %State{sock: sock, codec: codec}=state) do
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.push(0, encoded_payload))
+
+    {:noreply, state}
+  end
+
+  def handle_info({:tcp, socket, data}, %State{sock: socket}=state) do
+    make_active_once(state)
+    with {:ok, parsed_packets, leftover_data} <- Parser.parse(state.buffer, data) do
+      state = handle_packets(parsed_packets, state)
+
+      {:noreply, %State{state | buffer: leftover_data}}
+    else
+
+      {:error, {:need_more_data, data}} ->
+        {:noreply, %State{buffer: data }}
+    end
+  end
+
+  def handle_info({:tcp_closed, _socket}, _state) do
+    {:stop, :tcp_closed, nil}
+  end
+
+  def handle_info({:tcp_error, _, _}, _state) do
+    {:stop, :tcp_closed, nil}
+  end
+
+  def handle_info({:close_go_away, go_away_code, go_away_data}, %State{sequence: :go_away, waiters: waiters}=state) do
+    # tell the waiters that the remote end went away, then close.
+    err = {:error, {:remote_went_away, go_away_code, go_away_data}}
+    Enum.each(waiters, fn {_, waiter} ->
+      Connection.reply(waiter, err)
+    end)
+
+    {:disconnect, err, state}
+  end
+
+  def handle_info(:send_ping, %State{sock: sock, last_ping: nil}=state) do
+    {next_seq, state} = next_sequence(state)
+
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+    new_state = state
+      |> schedule_ping
+      |> Map.put(:last_ping, {next_seq, :erlang.system_time(:milli_seconds)})
+
+    {:noreply, new_state}
+  end
+  def handle_info(:send_ping, %State{last_ping: {sequence, _ts}}=state) do
+    {:stop, {:error, {:pong_not_received, sequence}}, state}
+  end
+
+  # Private
+
+  defp handle_packets([], state),
+    do: state
+  defp handle_packets([packet | rest], state) do
+    state = handle_packet(packet, state)
+    handle_packets(rest, state)
+  end
+
+  defp handle_packet({:ping, _flags, seq}, %State{sock: sock}=state) do
+    :gen_tcp.send(sock, Frames.pong(0, seq))
+    state
+  end
+
+  defp handle_packet({:pong, _flags, seq}, %State{last_ping: {seq, _ts}}=state) do
+    %State{state | last_ping: nil}
+  end
+
+  defp handle_packet({:pong, _flags, seq}, %State{last_ping: nil, waiters: waiters}=state) do
+    new_waiters =
+      case Map.pop(waiters, seq) do
+        {nil, waiters} ->
+          Logger.error("Got an unknown pong for sequence #{inspect seq}")
+          waiters
+        {reply_to, waiters} ->
+          Connection.reply(reply_to, :pong)
+          waiters
+      end
+
+    %State{state | waiters: new_waiters}
+  end
+
+  defp handle_packet({:hello_ack, _flags, ping_interval, data}, %State{}=state) do
+    [encoding, compression] = String.split(data, "|")
```

Comment:
Doing so will have to increment the version of the protocol, no? If there's something added, clients will need to account for it and not drop it on the ground. At least, I hope that's the case.